medcat/utils/ner/deid.py

```diff
@@ -62,18 +62,73 @@ def train(self, json_path: Union[str, list, None],
+
+        blocks = [[]]
+        if 'Roberta' in de_id_pipe.ner_pipe.tokenizer.__class__.__name__:
+
+            whitespace_pattern = re.compile(r'\s')
+            tok_output = de_id_pipe.ner_pipe.tokenizer(text, return_offsets_mapping=True)
+            for token, (start, end) in zip(tok_output['input_ids'], tok_output['offset_mapping']):
+                if token in [0, 2]:
+                    continue
+                if len(blocks[-1]) == maximum_tkns:
+                    to_append_block = []
+                    idx_chunk = -1
+                    for i in range(len(blocks[-1]) - 1, len(blocks[-1]) - 21, -1):
+                        # This method to avoid sub-word division is specific for RoBERTA's tokenizer (BPE)
+                        if re.search(whitespace_pattern, de_id_pipe.ner_pipe.tokenizer.decode(blocks[-1][i][0],clean_up_tokenization_spaces=False)):
+                            idx_chunk = i
+                            break
+
+                    if idx_chunk != -1:
+                        to_append_block.extend(blocks[-1][i:])
+                        del blocks[-1][idx_chunk:]
+
+                    to_append_block.append((token, (start, end)))
+                    blocks.append(to_append_block)
+
+                else:
+                    blocks[-1].append((token, (start, end)))
+
+        else:
+            # print("***********WARNING:************\nChunking functionality is implemented for RoBERTa model. The model detected is not RoBERTa, chunking omitted.\nPII information MAY BE REVEALED.")
+            warnings.warn("\n\nChunking functionality is implemented for RoBERTa models. The detected model is not RoBERTa, so chunking is omitted. Be cautious, as PII information MAY BE REVEALED.")
+            tok_output = de_id_pipe.ner_pipe.tokenizer(text, return_offsets_mapping=True)
+            for token, (start, end) in zip(tok_output['input_ids'], tok_output['offset_mapping']):
+                if token in [0, 2]:
+                    continue
+                blocks[-1].append((token, (start, end)))
+
+        return blocks
+
+    def deid_text(self, text: str, redact: bool = False,config: Optional[ConfigTransformersNER] = None) -> str:
         """Deidentify text and potentially redact information.
 
+        v2: Changed to address the limit of 512 tokens. Adding chunking (break down the document into mini-documents and then run the model)
+        v3 planned: Add overlapping window instead of a straight cut
+
         Args:
             text (str): The text to deidentify.
             redact (bool): Whether to redact the information.
 
         Returns:
             str: The deidentified text.
         """
-        self.cat.get_entities
-        return deid_text(self.cat, text, redact=redact)
+
+        if config is None:
+            config = ConfigTransformersNER()
+        maximum_tkns = min(510,config.general['maximum_tokens_model']) # This is specific to RoBERTa model, with a maximum token limit is 512.
```

Comment:
If we want to specify an in code maximum of 510, I'd prefer if this was in a module level constant (and probably in the new `chunking` module). That way we can change it during testing (i.e to run a test that does not do chunking which thus fails to de-id).